File: app/src/main/java/com/foxek/simpletimer/ui/timer/TimerPresenter.java
package com.foxek.simpletimer.ui.timer;

import com.foxek.simpletimer.R;
import com.foxek.simpletimer.ui.base.BasePresenter;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import static com.foxek.simpletimer.utils.Constants.TIMER_PLAYING;
import static com.foxek.simpletimer.utils.Constants.TIMER_STOPPED;
import static com.foxek.simpletimer.utils.IntervalUtils.formatIntervalData;
import static com.foxek.simpletimer.utils.IntervalUtils.formatIntervalNumber;

public class TimerPresenter extends BasePresenter<TimerContact.View> implements TimerContact.Presenter {

    TimerContact.Interactor interactor;

    private int currentInterval;
    @Inject
    public TimerPresenter(TimerContact.Interactor interactor) {
        this.interactor = interactor;
    }

    @Override
    public void detachView() {
        interactor.deleteDependencies();
        super.detachView();
    }

    @Override
    public void prepareIntervals(int workoutId) {

        getDisposable().add(interactor.getVolume(workoutId));


        getDisposable().add(interactor.fetchIntervalList(workoutId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(intervalSize -> {
                    interactor.setTimerState(TIMER_PLAYING);
                    getView().showCounterType(R.string.timer_prepare);
                    getView().showPlayInterface();
                    interactor.loadIntervalListToTimer();
                }, throwable -> {
                }));
    }

    @Override
    public void viewIsReady() {
        registerTimerCallback();
        registerTickCallback();
    }


    @Override
    public void pauseButtonClicked() {
        if (interactor.getTimerState()) {
            interactor.stopTimer();
            getView().showPauseInterface();
            interactor.setTimerState(TIMER_STOPPED);
        } else {
            interactor.continueTimer();
            getView().showPlayInterface();
            interactor.setTimerState(TIMER_PLAYING);
        }
    }

    @Override
    public void resetButtonClicked() {
        getView().startWorkoutActivity();
    }

    private void registerTimerCallback() {
        getDisposable().add(interactor.intervalFinishedCallback()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(intervalNumber -> {
                    if (intervalNumber < interactor.getIntervalSize()) {
                        interactor.indicateEndOfInterval();
                        if ((intervalNumber & 1) == 0)
                            getView().showCounterType(R.string.timer_rest_time);
                        else {
                            getView().showCounterType(R.string.timer_work_time);
                            getView().showCounterNumber(formatIntervalNumber(++currentInterval));
                            getView().showCounterName(interactor.getIntervalName(currentInterval));
                        }
                    } else {
                        interactor.indicateEndOfWorkout();
                        getView().startWorkoutActivity();
                    }
                }, throwable -> {
                })
        );
    }

    private void registerTickCallback() {
        getDisposable().add(interactor.tickCallback()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(time -> getView().showCurrentCounter(formatIntervalData(time)),
                        throwable -> {
                        }
                )
        );
    }

}

File: app/src/main/java/com/foxek/simpletimer/ui/timer/TimerPresenter.kt
package com.foxek.simpletimer.ui.timer

import com.foxek.simpletimer.R
import com.foxek.simpletimer.ui.base.BasePresenter
import com.foxek.simpletimer.utils.Constants.POST_TIME_TYPE
import com.foxek.simpletimer.utils.Constants.PREPARE_TIME_TYPE
import com.foxek.simpletimer.utils.Constants.REST_TIME_TYPE

import javax.inject.Inject

import io.reactivex.android.schedulers.AndroidSchedulers

import io.reactivex.schedulers.Schedulers

import com.foxek.simpletimer.utils.Constants.TIMER_PLAYING
import com.foxek.simpletimer.utils.Constants.TIMER_STOPPED
import com.foxek.simpletimer.utils.Constants.WORK_TIME_TYPE
import com.foxek.simpletimer.utils.formatIntervalData
import com.foxek.simpletimer.utils.formatIntervalNumber
import java.util.*

class TimerPresenter @Inject constructor(
        private var interactor: TimerContact.Interactor
) : BasePresenter<TimerContact.View>(), TimerContact.Presenter {

    private var currentInterval: Int = 0

    override fun detachView() {
        interactor.deleteDependencies()
        super.detachView()
    }

    override fun prepareIntervals(workoutId: Int) {

        disposable.add(interactor.getVolume(workoutId))

        disposable.add(interactor.fetchIntervalList(workoutId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    interactor.setTimerState(TIMER_PLAYING)
                    view?.showCounterType(R.string.timer_prepare)
                    view?.showPlayInterface()
                    interactor.loadIntervalListToTimer()
                }, {

                }))
    }

    override fun viewIsReady() {
        registerTimerCallback()
        registerTickCallback()
    }


    override fun pauseButtonClicked() {
        if (interactor.getTimerState()) {
            interactor.stopTimer()
            view?.showPauseInterface()
            interactor.setTimerState(TIMER_STOPPED)
        } else {
            interactor.continueTimer()
            view?.showPlayInterface()
            interactor.setTimerState(TIMER_PLAYING)
        }
    }

    override fun resetButtonClicked() {
        view?.startWorkoutActivity()
    }

    private fun registerTimerCallback() {
        disposable.add(interactor.intervalFinishedCallback()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ time ->
                    when (time.type) {
                        REST_TIME_TYPE -> {
                            interactor.indicateEndOfInterval()
                            view?.showCounterType(R.string.timer_rest_time)
                        }
                        WORK_TIME_TYPE -> {
                            interactor.indicateEndOfInterval()
                            view?.showCounterType(R.string.timer_work_time)
                            view?.showCounterNumber(formatIntervalNumber(++currentInterval))
                            view?.showCounterName(time.name)
                        }
                        POST_TIME_TYPE -> {
                            interactor.indicateEndOfWorkout()
                            view?.startWorkoutActivity()
                        }
                    }
                }, { })
        )
    }

    private fun registerTickCallback() {
        disposable.add(interactor.tickCallback()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ time -> view?.showCurrentCounter(formatIntervalData(time)) }, {})
        )
    }

}
